File: python/encapsulate_conditionals.py
# Encapsulate Conditionals
# Extract complex conditionals into functions that convey the intent of the condition.
# Create names that reveal the intent of the conditional.

# Before
class User:
    def __init__(self):
        self.age = 25
        self.children = False
        self.is_premium = False
        self.coupon = False

    def has_children(self):
        return self.children

    def premium_member(self):
        return self.is_premium

    def has_coupon(self):
        return self.coupon

def check_availability(user):
    if user.age > 18 and not user.has_children() and (user.premium_member() or user.has_coupon()):
        print("User is available for the offer.")
    else:
        print("User is not available for the offer.")

user = User()
check_availability(user)


# After
class User:
    def __init__(self):
        self.age = 25
        self.children = False
        self.is_premium = False
        self.coupon = False

    def has_children(self):
        return self.children

    def premium_member(self):
        return self.is_premium

    def has_coupon(self):
        return self.coupon

def check_availability(user):
    if _is_available(user):
        print("User is available for the offer.")
    else:
        print("User is not available for the offer.")

def _is_available(user):
    if user.age <= 18:
        return False
    if user.has_children():
        return False
    if user.premium_member():
        return True
    if user.has_coupon():
        return True
    return False

user = User()
check_availability(user)

File: javascript/encapsulate_conditionals.js
// Encapsulate Conditionals
// Extract complex conditionals into functions that convey the intent of the condition.
// Create names that reveal the intent of the conditional.

// Before
class User {
  constructor() {
    this.age = 25;
    this.children = false;
    this.isPremium = false;
    this.coupon = false;
  }

  hasChildren() {
    return this.children;
  }

  premiumMember() {
    return this.isPremium;
  }

  hasCoupon() {
    return this.coupon;
  }
}

function checkAvailability(user) {
  if (user.age > 18 && !user.hasChildren() && (user.premiumMember() || user.hasCoupon())) {
    console.log("User is available for the offer.");
  } else {
    console.log("User is not available for the offer.");
  }
}

const user = new User();
checkAvailability(user);


// After
class User {
  constructor() {
    this.age = 25;
    this.children = false;
    this.isPremium = false;
    this.coupon = false;
  }

  hasChildren() {
    return this.children;
  }

  premiumMember() {
    return this.isPremium;
  }

  hasCoupon() {
    return this.coupon;
  }
}

function checkAvailability(user) {
  if (_isAvailable(user)) {
    console.log("User is available for the offer.");
  } else {
    console.log("User is not available for the offer.");
  }
}

function _isAvailable(user) {
  if (user.age <= 18) {
    return false;
  }
  if (user.hasChildren()) {
    return false;
  }
  if (user.premiumMember()) {
    return true;
  }
  if (user.hasCoupon()) {
    return true;
  }
  return false;
}

const user = new User();
checkAvailability(user);
